aidial_adapter_bedrock/dial_api/storage.py

```diff
@@ -41,30 +63,85 @@ def to_form_data(
+        self,
+        directories: List[str],
+        filename: str,
+        content_type: str,
+        content: bytes,
     ) -> FileMetadata:
         async with aiohttp.ClientSession() as session:
-            data = FileStorage.to_form_data(filename, content_type, content)
-            async with session.post(
-                self.base_url,
+            data = FileStorage._to_form_data(filename, content_type, content)
+            async with session.put(
+                url="/".join([self.upload_url, *directories, filename]),
                 data=data,
-                headers=self.auth_headers(),
+                headers=self.auth.headers,
             ) as response:
                 response.raise_for_status()
                 meta = await response.json()
                 log.debug(
-                    f"Uploaded file: path={self.base_url}, file={filename}, metadata={meta}"
+                    f"Uploaded file: path={self.upload_url}, file={filename}, metadata={meta}"
                 )
                 return meta
 
+    async def upload_file_as_base64(
+        self, data: str, content_type: str
+    ) -> FileMetadata:
+        path_segments = _compute_path_as_hash(data)
+        directories, filename = path_segments[:-1], path_segments[-1]
+
+        ext = _get_extension(content_type)
+        filename = f"{filename}.{ext}" if ext is not None else filename
+
+        content: bytes = base64.b64decode(data)
+        return await self.upload(directories, filename, content_type, content)
+
 
-def _hash_digest(string: str) -> str:
-    return hashlib.sha256(string.encode()).hexdigest()
+def _compute_path_as_hash(file_content: str) -> List[str]:
+    hash = hashlib.sha256(file_content.encode()).hexdigest()
+    return _chunk_string([3, 3], hash)
 
 
-async def upload_base64_file(
-    storage: FileStorage, data: str, content_type: str
-) -> FileMetadata:
-    filename = _hash_digest(data)
-    content: bytes = base64.b64decode(data)
-    return await storage.upload(filename, content_type, content)
+def _chunk_string(chunk_sizes: List[int], string: str) -> List[str]:
+    chunks = []
+    start = 0
+    for chunk_size in chunk_sizes:
+        chunks.append(string[start : start + chunk_size])
+        start += chunk_size
+    chunks.append(string[start:])
+    return [chunk for chunk in chunks if chunk != ""]
+
+
+def _get_extension(content_type: str) -> Optional[str]:
+    pattern = r"^image\/(jpeg|jpg|png|bmp)$"
+    match = re.match(pattern, content_type)
+    return match.group(1) if match is not None else None
+
+
+DIAL_USE_FILE_STORAGE = get_env_bool("DIAL_USE_FILE_STORAGE", False)
+
+DIAL_URL: Optional[str] = None
+if DIAL_USE_FILE_STORAGE:
+    DIAL_URL = get_env(
+        "DIAL_URL", "DIAL_URL must be set to use the DIAL file storage"
+    )
+
+
+async def create_file_storage(
+    base_dir: str, headers: Mapping[str, str]
+) -> Optional[FileStorage]:
+    if not DIAL_USE_FILE_STORAGE or DIAL_URL is None:
```

Comment:
you can check only DIAL_USE_FILE_STORAGE because the case with DIAL_URL is already checked on initialization

Reply:
I wish. If I only check `DIAL_USE_FILE_STORAGE `, then the typechecker can't derive that `DIAL_URL` is string, it still thinks it's `Optional[str]`